syntax = "proto3";

package nvidia.riva.asr;

import "riva_audio.proto";

service RivaSpeechRecognition {
  rpc Recognize (RecognizeRequest) returns (RecognizeResponse) {}
  rpc StreamingRecognize (stream StreamingRecognizeRequest) returns (stream StreamingRecognizeResponse) {}
  rpc GetConfig (GetConfigRequest) returns (GetConfigResponse) {}
}

message RecognizeRequest {
  RecognitionConfig config = 1;
  bytes audio = 2;
}

message RecognizeResponse {
  repeated SpeechRecognitionResult results = 1;
}

message StreamingRecognizeRequest {
  oneof streaming_request {
    StreamingRecognizeConfig streaming_config = 1;
    bytes audio_content = 2;
  }
}

message StreamingRecognizeResponse {
  repeated SpeechRecognitionResult results = 1;
  SpeechRecognitionResult result = 2;
}

message StreamingRecognizeConfig {
  RecognitionConfig config = 1;
  bool interim_results = 2;
}

message GetConfigRequest {}

message GetConfigResponse {
  RecognitionConfig config = 1;
}

message RecognitionConfig {
  enum Encoding {
    ENCODING_UNSPECIFIED = 0;
    LINEAR_PCM = 1;
    FLAC = 2;
    MULAW = 3;
    ALAW = 4;
    OGGOPUS = 5;
  }

  Encoding encoding = 1;
  int32 sample_rate_hertz = 2;
  string language_code = 3;
  int32 max_alternatives = 4;
  bool profanity_filter = 5;
  // repeated SpeechContext speech_contexts = 6;
  int32 audio_channel_count = 7;
  bool enable_word_time_offsets = 8;
  // bool enable_character_time_offsets = 9; // Not in python list?
  // bool enable_word_confidence = 10; // Not in python list?
  bool enable_automatic_punctuation = 11;
  bool enable_separate_recognition_per_channel = 12;
  string model = 13;
  bool verbatim_transcripts = 14;
  // SpeakerDiarizationConfig diarization_config = 19;
  // map<string, string> custom_configuration = 24;
}

message SpeechRecognitionResult {
  repeated SpeechRecognitionAlternative alternatives = 1;
  bool is_final = 2;
  float stability = 3;
  int32 channel_tag = 4;
  int32 audio_processed = 5;
  int32 audio_received = 6;
  repeated WordInfo words = 7;
  int64 start_time = 8;
  int64 end_time = 9;
  int64 deleted_time = 10;
}

message SpeechRecognitionAlternative {
  string transcript = 1;
  float confidence = 2;
  repeated WordInfo words = 3;
}

message WordInfo {
  string word = 1;
  float confidence = 2;
  int32 start_time = 3;
  int32 end_time = 4;
  string speaker = 5;
}